From 8f4179ed6541308c2546a3832ba9f3a94b8f9b6f Mon Sep 17 00:00:00 2001
From: tommy-huang <tommy_huang@aspeedtech.com>
Date: Fri, 3 Dec 2021 14:42:32 +0800
Subject: [PATCH 04/10] drivers:i2c:Add to handle slave address received
 pending

Handle the to do case when salve address received pending
Refine re-trigger DMA case when issue occurred.

Signed-off-by: tommy-huang <tommy_huang@aspeedtech.com>
Change-Id: I3590cf0595443f40ce1d79d5b51aed2488d69f41
---
 drivers/i2c/i2c_aspeed.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

diff --git a/drivers/i2c/i2c_aspeed.c b/drivers/i2c/i2c_aspeed.c
index b8222da0b6..5cdfb789c0 100644
--- a/drivers/i2c/i2c_aspeed.c
+++ b/drivers/i2c/i2c_aspeed.c
@@ -176,6 +176,7 @@ LOG_MODULE_REGISTER(i2c_aspeed);
 #define AST_I2CS_ISR                    0x24
 
 #define AST_I2CS_ADDR_INDICAT_MASK      (3 << 30)
+#define AST_I2CS_Wait_DMA_Process   BIT(29)
 
 #define AST_I2CS_Wait_TX_DMA            BIT(25)
 #define AST_I2CS_Wait_RX_DMA            BIT(24)
@@ -1253,9 +1254,33 @@ int aspeed_i2c_slave_irq(const struct device *dev)
 	}
 
 	if (AST_I2CS_PKT_DONE & sts) {
-		sts &= ~(AST_I2CS_PKT_DONE | AST_I2CS_PKT_ERROR);
+		sts &= ~(AST_I2CS_PKT_DONE | AST_I2CS_PKT_ERROR | AST_I2CS_Wait_DMA_Process);
 		sys_write32(AST_I2CS_PKT_DONE, i2c_base + AST_I2CS_ISR);
+
 		switch (sts) {
+		case AST_I2CS_SLAVE_MATCH:
+			if (sys_read32(i2c_base + AST_I2CM_ISR)) {
+				LOG_DBG("S : Sw|D - Wait normal\n");
+			} else {
+				LOG_DBG("S : Sw|D - Issue rx dma\n");
+				if (slave_cb->write_requested) {
+					slave_cb->write_requested(data->slave_cfg);
+				}
+				cmd = AST_I2CS_ACTIVE_ALL | AST_I2CS_PKT_MODE_EN;
+				if (config->mode == DMA_MODE) {
+					cmd |= AST_I2CS_RX_DMA_EN;
+					sys_write32(AST_I2CS_SET_RX_DMA_LEN(I2C_SLAVE_BUF_SIZE)
+					, i2c_base + AST_I2CS_DMA_LEN);
+				} else if (config->mode == BUFF_MODE) {
+					cmd |= AST_I2CS_RX_BUFF_EN;
+					sys_write32(AST_I2CC_SET_RX_BUF_LEN(I2C_BUF_SIZE)
+					, i2c_base + AST_I2CC_BUFF_CTRL);
+				} else {
+					cmd &= ~AST_I2CS_PKT_MODE_EN;
+				}
+				sys_write32(cmd, i2c_base + AST_I2CS_CMD_STS);
+			}
+			break;
 		case AST_I2CS_SLAVE_MATCH | AST_I2CS_STOP:
 			LOG_DBG("S : Sw | P\n");
 			if (slave_cb->stop) {
-- 
2.17.1

